std/algorithm/iteration.d

```diff
@@ -5001,61 +5001,107 @@ private struct SplitterResult(alias isTerminator, Range)
 }
 
 /++
-Lazily splits the string `s` into words, using whitespace as the delimiter.
+Lazily splits the character-based range `s` into words, using whitespace as the
+delimiter.
 
-This function is string specific and, contrary to
+This function is character-range specific and, contrary to
 `splitter!(std.uni.isWhite)`, runs of whitespace will be merged together
 (no empty tokens will be produced).
 
 Params:
-    s = The string to be split.
+    s = The character-based range to be split. Must be a string, or a
+    random-access range of character types.
 
 Returns:
     An $(REF_ALTTEXT input range, isInputRange, std,range,primitives) of slices of
-    the original string split by whitespace.
+    the original range split by whitespace.
  +/
-auto splitter(C)(C[] s)
-if (isSomeChar!C)
+auto splitter(Range)(Range s)
+if (isSomeString!Range ||
+    isRandomAccessRange!Range && hasLength!Range && hasSlicing!Range &&
+    !isConvertibleToString!Range &&
+    isSomeChar!(ElementEncodingType!Range))
 {
     import std.algorithm.searching : find;
     static struct Result
     {
     private:
         import core.exception : RangeError;
-        C[] _s;
+        Range _s;
         size_t _frontLength;
 
-        void getFirst() pure @safe
+        void getFirst()
         {
             import std.uni : isWhite;
 
-            auto r = find!(isWhite)(_s);
-            _frontLength = _s.length - r.length;
+            static if (is(ElementType!Range == dchar) ||
+                       is(ElementType!Range == const(dchar)) ||
+                       is(ElementType!Range == immutable(dchar)))
+            {
+                // dchar range, we can just use find.
+                auto r = find!(isWhite)(_s.save);
+                _frontLength = _s.length - r.length;
+            }
+            else
+            {
+                // need to decode the characters until we find a space. This is
+                // ported from std.string.stripLeft.
```

Comment:
Is it really necessary to proliferate the evil that is auto-decoding?

>One of the worst things about autodecoding is it is special, it *only* steps in for strings. Fortunately, however, that specialness enabled us to save things with byCodePoint and byCodeUnit.

Reply:
This overload of `splitter` is special. It splits on *runs* of whitespace, which isn't possible with any of the other overloads. It's also one of the more used functions. That it doesn't work with arbitrary ranges of char is an artificial limitation.

In order for `byCodeUnit` to be useful in all places `string`s are useful, it has to have similar support from Phobos.

Besides, in this case, you are requesting auto-decoding, because you want to split on unicode whitespace.

Reply:
If your custom type is intended to behave like a string then like a string its `front` should return an entire `dchar` and its `popFront` should remove an entire `dchar`.

>In order for `byCodeUnit` to be useful in all places strings are useful, it has to have similar support from Phobos.

Isn't the entire point of `byCodeUnit` that it is a non-auto-decoding type whose elements are individual `char`s and _not_ `dchar`s?

Reply:
> Isn't the entire point of `byCodeUnit` that it is a non-auto-decoding type whose elements are individual `char`s and _not_ `dchar`s?

Yes, but that doesn't mean you can't use it to decode `dchar`s.

Consider that if you have a `byCodeUnit` range, `bcu = strip(bcu)` still works and strips unicode spaces. Why shouldn't `splitter` enjoy the same benefits?

Autodecoding doesn't mean NO decoding, it just means I have to opt in to use it. Using this form of `splitter` is opting in.

> If your custom type is intended to behave like a string then like a string its `front` should return an entire `dchar` and its `popFront` should remove an entire `dchar`.

It should be `string`-like in all respects _except_ this one. That is, I want to use it as an array, but treat it like a string for printing, processing, using with `strip`, `splitter`, etc.

Reply:
> in this case, you are requesting auto-decoding

I said this wrong. In this case you are requesting _decoding_.

Reply:
> What about giving it an inverse of `byCodeUnit` and explicitly opting in to decoding, like the reverse is done for strings?

No, because what I am ultimately developing (not here, but this is a requirement) is that `splitter` returns the *exact* range type I pass in. If I have to change to another range type, then I lose the extra information I've added to my specialized range.

It's also an odd extra step, when the guts of this function, `stripLeft`, work fine on a `char`-based range.

> If I want my range _not_ to be cross-element decoded what should I use to split on runs of consecutive whitespace?

I'm not understanding this. The elements returned by `splitter` will not turn into auto-decoded ranges, so how does this not satisfy your requirements?

Reply:
I don't mean the return type, I mean the performance cost.

Reply:
I suppose the right question to answer is, how do I check for whitespace using `std.uni.isWhite` without decoding?

Reply:
Since `std.string.stripLeft` works like this already when given non-string ranges I'm coming to believe this might be correct.